rclcpp/src/rclcpp/node_interfaces/node_graph.cpp

```diff
@@ -178,6 +179,136 @@ NodeGraph::get_service_names_and_types_by_node(
   return services_and_types;
 }
 
+std::map<std::string, std::vector<std::string>>
+NodeGraph::get_client_names_and_types_by_node(
+  const std::string & node_name,
+  const std::string & namespace_) const
+{
+  rcl_names_and_types_t service_names_and_types = rcl_get_zero_initialized_names_and_types();
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcl_ret_t ret = rcl_get_client_names_and_types_by_node(
+    node_base_->get_rcl_node_handle(),
+    &allocator,
+    node_name.c_str(),
+    namespace_.c_str(),
+    &service_names_and_types);
+  if (ret != RCL_RET_OK) {
+    auto error_msg = std::string("failed to get service names and types by node: ") +
+      rcl_get_error_string().str;
+    rcl_reset_error();
+    if (rcl_names_and_types_fini(&service_names_and_types) != RCL_RET_OK) {
+      error_msg +=
+        std::string(", failed also to cleanup service names and types, leaking memory: ") +
+        rcl_get_error_string().str;
+      rcl_reset_error();
+    }
+    throw std::runtime_error(error_msg);
+  }
```

Comment:
Maybe using [throw_from_rcl_error](https://github.com/ros2/rclcpp/blob/63e79223f98867ba00c059eb612438ca9461a2f7/rclcpp/include/rclcpp/exceptions/exceptions.hpp#L128) and [rcpputils::make_scope_exit](https://github.com/ros2/rcpputils/blob/bc300e7667603a4b8a6cc2b9d7b13f5b64e79265/include/rcpputils/scope_exit.hpp#L58) (?)

something like:

```cpp
  rcl_names_and_types_t service_names_and_types = rcl_get_zero_initialized_names_and_types();
  auto service_names_and_types_finalizer = rcpputils::make_scope_exit([&service_names_and_types] () {
    if (rcl_names_and_types_fini(&service_names_and_types) != RCL_RET_OK) {
      RCLCPP_ERROR(....);
      rcl_reset_error();
    }
  });
  rcl_ret_t ret = rcl_get_client_names_and_types_by_node(...);
  if (ret != RCL_RET_OK) {
    rclcpp::exception::throw_from_rcl_error(...);
  }
  ...
  service_names_and_types_finalizer.cancel();
```

Reply:
similar comment for the other implementations

Reply:
i think that would be nicer. borrowing your pseudo code.